### ISSUE TITLE
FedAvg results

### ISSUE THREAD

**Issue by cnyanhao** [YamingGuo98/FedIIR#3]
Hi, thanks for your great work and neat codebase on federated domain generalization.

I run the code for FedAvg with the following commands.

```
python main.py --dataset RotatedMNIST --algorithm FedAvg --num_clients 50 --output_dir train_output --skip_model_save
python main.py --dataset VLCS --algorithm FedAvg --num_clients 50 --output_dir train_output --skip_model_save
python main.py --dataset PACS --algorithm FedAvg --num_clients 50 --output_dir train_output --skip_model_save
python main.py --dataset OfficeHome --algorithm FedAvg --num_clients 50 --output_dir train_output --skip_model_save
```

I only run the experiment once for each dataset and got the following results.

```
Dataset,Algorithm,Num_clients,Trial_seed,Seed,lr,batch_size,Test_env,Train_acc,Test_acc
RotatedMNIST,FedAvg,50,0,0,0.01,64,Avg,0.965809033733562,0.9328566598765788
VLCS,FedAvg,50,0,0,0.001,32,Avg,0.8326586794621905,0.7317361037045496
PACS,FedAvg,50,0,0,0.0025,32,Avg,0.9086598468140008,0.7547587749368678
OfficeHome,FedAvg,50,0,0,0.001,32,Avg,0.7858567516530083,0.6559344883918246
```

It seems that the results are way better than the FedAvg results reported in the paper. This is shown in the following table.

|  | RotatedMNIST | VLCS | PACS | OfficeHome |
|---|---|---|---|---|
| FedAvg (Code) | 93 | 73 | 75 | 66 |
| FedAvg (Paper) | 91 | 65 | 69 | 60 |
| FedIIR (Paper) | 93 | 74 | 75 | 66 |

Note that I used Python 3.10.13, PyTorch 2.1.1 and TorchVision 0.16.1. These are different from your environment.

May I kindly ask possible reasons for such large difference? Or did I make some thing wrong, like some hyper-parameters? Do you think the library version will lead to such difference?

Looking forward to your reply. Thank you in advance.

**Comment by ymguo21** [comment]
I sincerely apologize for the delayed response, as I missed the email from GitHub. Thank you for your attention and interest in our work. I understand your concerns about the observed differences in results compared to those reported in the paper. Allow me to address your inquiries and provide some insights:

1. I would like to clarify that, for each algorithm and dataset, the learning rate was setted through grid search. Due to time and lab's resources constraints, hyperparameters were only tuned in a small number of clients scenario, and they remained constant across all the scenarios with a large number of clients. The learning rates used in our experiments are as follows:

      |    | RotatedMNIST   | VLCS   | PACS   | OfficeHome   |
      |-------|-------|-------|-------|-------|
      | FedAvg/FedAdg  | 0.005 | 0.0001 | 0.001 | 0.00025  |
      | FedSR/FedIIR   | 0.01 | 0.001  | 0.0025  | 0.001 |



2. As you rightly pointed out, differences in the library versions can indeed impact results, especially in scenarios with a large number of clients. Large client scenarios involve more complex data splitting and client sampling, and even minor variations can lead to significant differences in model performance.

If you are interested in scenarios with a large number of clients, I recommend trying to individually tune hyperparameters for each setting to achieve better results. 
Once again, thank you for your attention and understanding!
